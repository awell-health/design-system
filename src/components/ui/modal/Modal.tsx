import React, { ReactNode, type FC, type RefObject } from 'react';
import { useClickAway, useKey } from 'react-use';
import { cn } from '../../../lib/utils';

interface Props {
  onClose: () => void;
  clickOutside?: boolean;
  closeOnEscape?: boolean;
  children: JSX.Element | ReactNode | string;
  className?: string;
  marginTop?: string;
}

const Modal: FC<Props> = (props) => {
  const {
    children,
    onClose,
    clickOutside = false,
    closeOnEscape = false,
    className = '',
    marginTop = 'mt-[100px]'
  } = props;
  const modalRef: RefObject<HTMLDivElement> = React.useRef<HTMLDivElement>(null);

  if (clickOutside) {
    useClickAway(modalRef, onClose);
  }

  if (closeOnEscape) {
    useKey('Escape', onClose);
  }

  return (
    <div
      className='relative z-[9999]'
      aria-labelledby='modal-title'
      role='dialog'
      aria-modal='true'
    >
      <div className='fixed inset-0 bg-slate-700 bg-opacity-30 transition-opacity z-[9999]'></div>
      <div className='fixed inset-0 z-[9999] w-screen overflow-y-auto'>
        <div className={cn('flex justify-center p-4', marginTop)}>
          <div className='relative transform overflow-hidden transition-all bg-white rounded-lg shadow border border-slate-200'>
            <div ref={modalRef} className={cn('w-[1024px] h-[500px]', className)}>
              {children}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export { Modal };
