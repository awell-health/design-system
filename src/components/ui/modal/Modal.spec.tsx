import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';

import { Modal } from './Modal';
import React from 'react';

const handleClose = vi.fn();

vi.mock('react-use', () => ({
  useClickAway: vi.fn().mockImplementation(() => {
    handleClose();
  }),
  useKey: vi.fn().mockImplementation(() => {
    handleClose();
  })
}));

describe('Modal', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      children: 'Modal Content',
      onClose: handleClose
    };

    return render(<Modal {...defaultProps} {...props} />);
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('match snapshot', () => {
    expect(subject().container).toMatchSnapshot();
  });

  it('triggers handleClose function', () => {
    const children = <div onClick={handleClose}>handle close</div>;
    const { getByText } = subject({ children });

    fireEvent.click(getByText('handle close'));

    expect(handleClose).toHaveBeenCalledOnce();
  });

  it('triggers handleClose on Escape key press', () => {
    const children = <div onClick={handleClose}>handle close</div>;
    const { container } = subject({ children, closeOnEscape: true });

    fireEvent.keyDown(container, 'Escape');

    expect(handleClose).toHaveBeenCalledOnce();
  });

  it('triggers handleClose on click away', () => {
    const children = 'Test';
    const { getByRole } = subject({ children, clickOutside: true });

    fireEvent.click(getByRole('dialog'));

    expect(handleClose).toHaveBeenCalledOnce();
  });
});
