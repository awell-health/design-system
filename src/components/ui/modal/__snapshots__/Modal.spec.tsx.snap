// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > match snapshot 1`] = `
<div>
  <div
    aria-labelledby="modal-title"
    aria-modal="true"
    class="relative z-[9999]"
    role="dialog"
  >
    <div
      class="fixed inset-0 bg-slate-700 bg-opacity-30 transition-opacity z-[9999]"
    />
    <div
      class="fixed inset-0 z-[9999] w-screen overflow-y-auto"
    >
      <div
        class="flex justify-center p-4 mt-[100px]"
      >
        <div
          class="relative transform transition-all bg-white rounded-lg shadow border border-slate-200"
        >
          <div
            class="w-[1024px] h-[500px]"
          >
            Modal Content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
